sentry-core/src/session.rs

```diff
@@ -293,11 +397,94 @@ mod tests {
         let items = envelopes[0].items().chain(envelopes[1].items());
         assert_eq!(items.clone().count(), MAX_SESSION_ITEMS * 2);
         for item in items {
-            assert!(match item {
-                EnvelopeItem::SessionUpdate(_) => true,
-                _ => false,
+            assert!(matches!(item, EnvelopeItem::SessionUpdate(_)));
+        }
+    }
+
+    #[test]
+    fn test_session_aggregation() {
+        let envelopes = crate::test::with_captured_envelopes_options(
+            || {
+                sentry::start_session();
+                // this error will be captured along with an individual update.
+                let err = "NaN".parse::<usize>().unwrap_err();
+                sentry::capture_error(&err);
+
+                for _ in 0..50 {
+                    sentry::start_session();
+                }
+                sentry::end_session();
+
+                sentry::configure_scope(|scope| {
+                    scope.set_user(Some(sentry::User {
+                        id: Some("foo-bar".into()),
+                        ..Default::default()
+                    }));
+                    scope.add_event_processor(Box::new(|_| None));
+                });
+
+                for _ in 0..50 {
+                    sentry::start_session();
+                }
+
+                // this error will be discarded because of the event processor,
+                // but the session will still be updated accordingly.
+                let err = "NaN".parse::<usize>().unwrap_err();
+                sentry::capture_error(&err);
+            },
+            crate::ClientOptions {
+                release: Some("some-release".into()),
+                session_mode: SessionMode::Request,
+                ..Default::default()
+            },
+        );
+        assert_eq!(envelopes.len(), 3);
+
+        let session_id;
+
+        let mut items = envelopes[0].items();
+        assert!(matches!(items.next(), Some(EnvelopeItem::Event(_))));
+        if let Some(EnvelopeItem::SessionUpdate(session)) = items.next() {
+            session_id = session.session_id;
+            assert_eq!(session.status, SessionStatus::Ok);
+            assert_eq!(session.errors, 1);
+            assert_eq!(session.init, true);
+        } else {
+            panic!("expected session");
+        }
```

Comment:
`items.next().expect("expected session")` would be more concise?

Reply:
well no, since the next item can also be an event.